nft-marketplace/marketplace/src/auction.rs

```diff
@@ -0,0 +1,218 @@
+    /// * `token_id`: the NFT id
+    /// * `min_price`: the starting price
+    /// * `bid_period`: the time that the auction lasts until another bid occurs
+    pub async fn create_auction(
+        &mut self,
+        nft_contract_id: &ActorId,
+        ft_contract_id: Option<ActorId>,
+        token_id: U256,
+        min_price: u128,
+        bid_period: u64,
+        duration: u64,
+    ) {
+        self.check_approved_nft_contract(nft_contract_id);
+        self.check_approved_ft_contract(ft_contract_id);
+        let contract_and_token_id =
+            format!("{}{}", H256::from_slice(nft_contract_id.as_ref()), token_id);
+        self.on_auction(&contract_and_token_id);
+        if bid_period < MIN_BID_PERIOD || duration < MIN_BID_PERIOD {
+            panic!("bid period or auction duration can't be less than 1 minute");
+        }
+        if min_price == 0 {
+            panic!("price can't be equal to zero");
+        }
+        // approve nft to trade on the marketplace
+        nft_approve(nft_contract_id, &exec::program_id(), token_id).await;
+
+        let auction = Auction {
+            bid_period,
+            started_at: exec::block_timestamp(),
+            ended_at: exec::block_timestamp() + duration,
+            current_price: min_price,
+            bids: None,
+        };
+        self.items
+            .entry(contract_and_token_id)
+            .and_modify(|item| {
+                item.price = None;
+                item.auction = Some(auction.clone());
+                item.ft_contract_id = ft_contract_id
+            })
+            .or_insert(Item {
+                owner_id: msg::source(),
+                nft_contract_id: *nft_contract_id,
+                ft_contract_id,
+                token_id,
+                price: None,
+                auction: Some(auction),
+                offers: Vec::new(),
+            });
+
+        msg::reply(
+            MarketEvent::AuctionCreated {
+                nft_contract_id: *nft_contract_id,
+                token_id,
+                price: min_price,
+            },
+            0,
+        );
+    }
+
+    /// Settles the auction
+    /// Requirements:
+    /// * The auction must be over
+    /// Arguments:
+    /// * `nft_contract_id`: the NFT contract address
+    /// * `token_id`: the NFT id
+    pub async fn settle_auction(&mut self, nft_contract_id: &ActorId, token_id: U256) {
+        let contract_and_token_id =
+            format!("{}{}", H256::from_slice(nft_contract_id.as_ref()), token_id);
+        let item = self
+            .items
+            .get_mut(&contract_and_token_id)
+            .expect("Item does not exist");
+
+        let auction = item.auction.clone().expect("Auction doesn not exist");
+
+        if auction.ended_at > exec::block_timestamp() {
+            panic!("Auction is not over");
+        }
+
+        let mut bids = auction.bids.unwrap_or_default();
+
+        if !bids.is_empty() {
```

Comment:
This if also could be rewritten to deacrease nesting and make the code more consequent. I'm not insisting though =)
```suggestion
        let mut bids = auction.bids.unwrap_or_default();
        if bids.is_empty() {
             msg::reply(
                MarketEvent::AuctionCancelled {
                    nft_contract_id: *nft_contract_id,
                    token_id,
                },
                0,
            );
            
            return;
        }
        
        let highest_bid = &bids[bids.len() - 1].clone();
        bids.pop();
        for bid in bids {
            transfer_payment(&exec::program_id(), &bid.id, item.ft_contract_id, bid.price)
                .await;
        }
        // fee for treasury
        // ...
```